File: backend/src/hearthd/state.zig
//! In-memory control-plane state with a mutex and JSON-file persistence.
//! All string fields are owned by `gpa` (duplicated on insert).
const std = @import("std");
const Io = std.Io;
const common = @import("common");
const models = common.models;
const jsonh = common.jsonh;

pub const State = struct {
    gpa: std.mem.Allocator,
    mu: common.SpinLock = .{},
    nodes: std.ArrayList(models.Node) = .empty,
    sandboxes: std.ArrayList(models.Sandbox) = .empty,
    request_count: u64 = 0,
    seq: u64 = 0,
    // Wake/fork metrics (API-V2.md §7).
    wake_ms_last: u64 = 0,
    wake_total: u64 = 0,
    wake_ms_sum: u64 = 0,
    forks_total: u64 = 0,
    prng: std.Random.DefaultPrng = std.Random.DefaultPrng.init(0),

    pub fn init(gpa: std.mem.Allocator) State {
        var s = State{ .gpa = gpa };
        // Seed from the address of a stack variable for per-run uniqueness.
        var anchor: u8 = 0;
        s.prng = std.Random.DefaultPrng.init(@intFromPtr(&anchor));
        return s;
    }

    pub fn bumpRequests(self: *State) void {
        self.mu.lock();
        defer self.mu.unlock();
        self.request_count += 1;
    }

    // ---- ids ----
    fn nextId(self: *State, prefix: []const u8) ![]const u8 {
        self.seq += 1;
        // Ids need only be unique, not cryptographically random.
        const r = self.prng.random().int(u32);
        return std.fmt.allocPrint(self.gpa, "{s}-{x:0>8}-{d}", .{ prefix, r, self.seq });
    }

    // ---- nodes ----

    pub fn findNode(self: *State, id: []const u8) ?*models.Node {
        for (self.nodes.items) |*n| {
            if (std.mem.eql(u8, n.id, id)) return n;
        }
        return null;
    }

    fn findNodeByHostname(self: *State, hostname: []const u8) ?*models.Node {
        for (self.nodes.items) |*n| {
            if (std.mem.eql(u8, n.hostname, hostname)) return n;
        }
        return null;
    }

    /// Register (idempotent by hostname). Returns the node id (owned by state).
    pub fn registerNode(self: *State, hostname: []const u8, addr: []const u8, cpus: u32, mem_total: u64, now: i64) ![]const u8 {
        self.mu.lock();
        defer self.mu.unlock();

        if (self.findNodeByHostname(hostname)) |n| {
            // Update mutable fields, keep id.
            self.gpa.free(n.addr);
            n.addr = try self.gpa.dupe(u8, addr);
            n.cpus = cpus;
            n.mem_total_mib = mem_total;
            n.last_heartbeat = now;
            return n.id;
        }
        const id = try self.nextId("node");
        const node = models.Node{
            .id = id,
            .hostname = try self.gpa.dupe(u8, hostname),
            .addr = try self.gpa.dupe(u8, addr),
            .cpus = cpus,
            .mem_total_mib = mem_total,
            .mem_free_mib = mem_total,
            .vm_count = 0,
            .last_heartbeat = now,
        };
        try self.nodes.append(self.gpa, node);
        return id;
    }

    pub fn heartbeat(self: *State, id: []const u8, mem_free: u64, vm_count: u32, pool_size: u32, now: i64) bool {
        self.mu.lock();
        defer self.mu.unlock();
        const n = self.findNode(id) orelse return false;
        n.mem_free_mib = mem_free;
        n.vm_count = vm_count;
        n.pool_size = pool_size;
        n.last_heartbeat = now;
        return true;
    }

    /// Pick the ready node with the lowest vm_count. Caller must hold the lock.
    pub fn pickNode(self: *State, now: i64) ?*models.Node {
        var best: ?*models.Node = null;
        for (self.nodes.items) |*n| {
            if (n.status(now) != .ready) continue;
            if (best == null or n.vm_count < best.?.vm_count) best = n;
        }
        return best;
    }

    // ---- sandboxes ----

    pub fn findSandbox(self: *State, id: []const u8) ?*models.Sandbox {
        for (self.sandboxes.items) |*s| {
            if (std.mem.eql(u8, s.id, id)) return s;
        }
        return null;
    }

    /// Create a sandbox record in "creating" state. Caller must hold the lock.
    pub fn createSandbox(self: *State, name: []const u8, namespace: []const u8, node_id: []const u8, vcpus: u32, mem_mib: u64, now: i64) !*models.Sandbox {
        const id = try self.nextId("sb");
        const sb = models.Sandbox{
            .id = id,
            .name = try self.gpa.dupe(u8, name),
            .namespace = try self.gpa.dupe(u8, namespace),
            .node_id = try self.gpa.dupe(u8, node_id),
            .state = .creating,
            .vcpus = vcpus,
            .mem_mib = mem_mib,
            .ip = null,
            .created_at = now,
            .parent_id = null,
        };
        try self.sandboxes.append(self.gpa, sb);
        return &self.sandboxes.items[self.sandboxes.items.len - 1];
    }

    pub fn setSandboxState(self: *State, id: []const u8, st: models.SandboxState) void {
        self.mu.lock();
        defer self.mu.unlock();
        if (self.findSandbox(id)) |s| s.state = st;
    }

    /// Update the `ip` field of a sandbox (duping the new value). A null clears.
    pub fn setSandboxIp(self: *State, id: []const u8, ip: ?[]const u8) void {
        self.mu.lock();
        defer self.mu.unlock();
        const s = self.findSandbox(id) orelse return;
        if (s.ip) |old| self.gpa.free(old);
        s.ip = if (ip) |v| (self.gpa.dupe(u8, v) catch null) else null;
    }

    pub fn recordWake(self: *State, ms: u64) void {
        self.mu.lock();
        defer self.mu.unlock();
        self.wake_ms_last = ms;
        self.wake_total += 1;
        self.wake_ms_sum += ms;
    }

    pub fn recordFork(self: *State) void {
        self.mu.lock();
        defer self.mu.unlock();
        self.forks_total += 1;
    }

    /// Create a child sandbox record (fork). Caller must hold the lock.
    pub fn createForkChild(
        self: *State,
        name: []const u8,
        namespace: []const u8,
        node_id: []const u8,
        vcpus: u32,
        mem_mib: u64,
        parent_id: []const u8,
        now: i64,
    ) !*models.Sandbox {
        const id = try self.nextId("sb");
        const sb = models.Sandbox{
            .id = id,
            .name = try self.gpa.dupe(u8, name),
            .namespace = try self.gpa.dupe(u8, namespace),
            .node_id = try self.gpa.dupe(u8, node_id),
            .state = .creating,
            .vcpus = vcpus,
            .mem_mib = mem_mib,
            .ip = null,
            .created_at = now,
            .parent_id = try self.gpa.dupe(u8, parent_id),
        };
        try self.sandboxes.append(self.gpa, sb);
        return &self.sandboxes.items[self.sandboxes.items.len - 1];
    }

    pub fn removeSandbox(self: *State, id: []const u8) void {
        self.mu.lock();
        defer self.mu.unlock();
        var i: usize = 0;
        while (i < self.sandboxes.items.len) : (i += 1) {
            if (std.mem.eql(u8, self.sandboxes.items[i].id, id)) {
                const s = self.sandboxes.orderedRemove(i);
                self.freeSandbox(s);
                return;
            }
        }
    }

    fn freeSandbox(self: *State, s: models.Sandbox) void {
        self.gpa.free(s.id);
        self.gpa.free(s.name);
        self.gpa.free(s.namespace);
        if (s.node_id) |v| self.gpa.free(v);
        if (s.ip) |v| self.gpa.free(v);
        if (s.parent_id) |v| self.gpa.free(v);
    }

    // ---- persistence ----

    /// Serialize the full state to JSON and atomically write it (tmp + rename).
    pub fn persist(self: *State, io: Io, path: []const u8) !void {
        self.mu.lock();
        defer self.mu.unlock();

        var arena_state = std.heap.ArenaAllocator.init(self.gpa);
        defer arena_state.deinit();
        const arena = arena_state.allocator();

        var b = jsonh.Builder.init(arena);
        try b.raw("{\"seq\":");
        try b.print("{d}", .{self.seq});
        try b.raw(",\"request_count\":");
        try b.print("{d}", .{self.request_count});
        try b.raw(",\"nodes\":[");
        var first = true;
        for (self.nodes.items) |n| {
            if (!first) try b.byte(',');
            first = false;
            try n.writeJson(&b, n.last_heartbeat); // status recomputed on load
        }
        try b.raw("],\"sandboxes\":[");
        first = true;
        for (self.sandboxes.items) |s| {
            if (!first) try b.byte(',');
            first = false;
            try s.writeJson(&b);
        }
        try b.raw("]}");

        const tmp = try std.fmt.allocPrint(arena, "{s}.tmp", .{path});
        const dir = Io.Dir.cwd();
        try dir.writeFile(io, .{ .sub_path = tmp, .data = b.items() });
        try dir.rename(tmp, dir, path, io);
    }

    /// Load state from a JSON file (best effort). Strings duped into gpa.
    pub fn load(self: *State, io: Io, path: []const u8) !void {
        self.mu.lock();
        defer self.mu.unlock();

        const dir = Io.Dir.cwd();
        const data = try dir.readFileAlloc(io, path, self.gpa, .limited(16 * 1024 * 1024));
        defer self.gpa.free(data);

        var parsed = try jsonh.parse(self.gpa, data);
        defer parsed.deinit();
        const root = parsed.root();

        if (jsonh.getInt(root, "seq")) |v| self.seq = @intCast(v);
        if (jsonh.getInt(root, "request_count")) |v| self.request_count = @intCast(@max(0, v));

        if (jsonh.get(root, "nodes")) |nodes_v| {
            if (nodes_v == .array) {
                for (nodes_v.array.items) |nv| {
                    const n = models.Node{
                        .id = try self.gpa.dupe(u8, jsonh.getString(nv, "id") orelse continue),
                        .hostname = try self.gpa.dupe(u8, jsonh.getString(nv, "hostname") orelse ""),
                        .addr = try self.gpa.dupe(u8, jsonh.getString(nv, "addr") orelse ""),
                        .cpus = @intCast(jsonh.getInt(nv, "cpus") orelse 0),
                        .mem_total_mib = @intCast(jsonh.getInt(nv, "mem_total_mib") orelse 0),
                        .mem_free_mib = @intCast(jsonh.getInt(nv, "mem_free_mib") orelse 0),
                        .vm_count = @intCast(jsonh.getInt(nv, "vm_count") orelse 0),
                        .pool_size = @intCast(jsonh.getInt(nv, "pool_size") orelse 0),
                        .last_heartbeat = jsonh.getInt(nv, "last_heartbeat") orelse 0,
                    };
                    try self.nodes.append(self.gpa, n);
                }
            }
        }
        if (jsonh.get(root, "sandboxes")) |sv| {
            if (sv == .array) {
                for (sv.array.items) |s| {
                    const state_str = jsonh.getString(s, "state") orelse "stopped";
                    const sb = models.Sandbox{
                        .id = try self.gpa.dupe(u8, jsonh.getString(s, "id") orelse continue),
                        .name = try self.gpa.dupe(u8, jsonh.getString(s, "name") orelse ""),
                        .namespace = try self.gpa.dupe(u8, jsonh.getString(s, "namespace") orelse "default"),
                        .node_id = if (jsonh.getString(s, "node_id")) |v| try self.gpa.dupe(u8, v) else null,
                        .state = models.SandboxState.fromString(state_str) orelse .stopped,
                        .vcpus = @intCast(jsonh.getInt(s, "vcpus") orelse 1),
                        .mem_mib = @intCast(jsonh.getInt(s, "mem_mib") orelse 256),
                        .ip = if (jsonh.getString(s, "ip")) |v| try self.gpa.dupe(u8, v) else null,
                        .created_at = jsonh.getInt(s, "created_at") orelse 0,
                        .parent_id = if (jsonh.getString(s, "parent_id")) |v| try self.gpa.dupe(u8, v) else null,
                    };
                    try self.sandboxes.append(self.gpa, sb);
                }
            }
        }
    }
};

File: go/internal/state/state.go
// Package state implements the in-memory control-plane state store with
// JSON-file persistence, replicating state.zig exactly.
package state

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sync"

	"github.com/alpham/infra-saas/hearth/internal/model"
)

// State is the central in-memory store. All exported methods are safe for
// concurrent use via an internal mutex.
type State struct {
	mu sync.Mutex

	Nodes     []*model.Node
	Sandboxes []*model.Sandbox

	RequestCount uint64
	Seq          uint64

	// Wake/fork/exec metrics.
	WakeMsLast uint64
	WakeTotal  uint64
	WakeMsSum  uint64
	ForksTotal uint64
	ExecsTotal uint64
}

// New creates an empty State.
func New() *State {
	return &State{}
}

// Lock acquires the state mutex. Use for multi-step read/write sequences.
func (s *State) Lock() { s.mu.Lock() }

// Unlock releases the state mutex.
func (s *State) Unlock() { s.mu.Unlock() }

// BumpRequests increments the request counter (called on every inbound request).
func (s *State) BumpRequests() {
	s.mu.Lock()
	s.RequestCount++
	s.mu.Unlock()
}

// ---- ID generation ----

// IDRandomBytes is the crypto/rand width of the random half of every id.
// A sandbox id lands in the public ingress label "<expose-name>--<id>"
// (ADR-0007), which is a single DNS label capped at 63 octets: 32 for the
// longest expose name plus the 2-byte separator leaves 29 for the id, i.e.
// "sb-" and 26 hex chars. That is 104 bits — enough that the label is a real
// bearer capability, which it must be because the gateway authenticates
// nothing on an inbound ingress request.
const IDRandomBytes = 13

// randomHex returns n cryptographically random bytes hex-encoded.
func randomHex(n int) string {
	b := make([]byte, n)
	if _, err := rand.Read(b); err != nil {
		// crypto/rand failure is unrecoverable for id issuance.
		panic(fmt.Sprintf("crypto/rand: %v", err))
	}
	return hex.EncodeToString(b)
}

// nextID generates a new ID with the given prefix.
// Format: "<prefix>-<26 hex chars>". Seq still advances (it is part of the
// persisted state format) but stays out of the id: a monotonic tail would
// hand an attacker the position of every other tenant's id in the stream,
// and it costs label budget that entropy needs more.
// Caller must hold the lock.
func (s *State) nextID(prefix string) string {
	s.Seq++
	return prefix + "-" + randomHex(IDRandomBytes)
}

// ---- Nodes ----

// FindNode returns the node with the given id, or nil.
// Caller must hold the lock.
func (s *State) FindNode(id string) *model.Node {
	for _, n := range s.Nodes {
		if n.ID == id {
			return n
		}
	}
	return nil
}

func (s *State) findNodeByHostname(hostname string) *model.Node {
	for _, n := range s.Nodes {
		if n.Hostname == hostname {
			return n
		}
	}
	return nil
}

// MaxNodes caps how many node records the working set may hold. Every node is
// scanned linearly under this mutex by FindNode/PickNode/the registration
// checks, and a full SaveSnapshot rewrites all of them on every state change —
// so the list must not be able to grow without bound. The cap is far above the
// ADR-0002 single-hearthd scale point, so it is a backstop and not a scheduling
// limit; the caller-supplied admission check (see RegisterNodeGuarded) is what
// actually closes the growth path.
const MaxNodes = 1024

// StaleNodeTTL is how long a node record must have been silent before
// RegisterNodeGuarded may reclaim it to make room at the MaxNodes ceiling.
//
// A hard ceiling with no reclamation path is its own outage: records left by
// re-enrollments under fresh wireguard keys accumulate, and once the table is
// full every genuinely new node gets a 503 forever — while the agent
// (rust/agent/src/registration.rs) retries register every 5s and never reaches
// its heartbeat loop. So the ceiling needs a way down, and reclaiming a
// LEGITIMATE node's record would be worse than the growth: it drops the address
// hearthd dials that worker's VMs at.
//
// Seven days is therefore deliberately enormous. An agent heartbeats every 5s
// and a node is "down" after 15s of silence (model.Node.NodeStatus), so this is
// ~120,000 consecutive missed heartbeats: a reboot, a long maintenance window, a
// network partition, or a hearthd restart from a slightly stale snapshot all sit
// many orders of magnitude below it. Nothing that is coming back is reclaimed.
const StaleNodeTTL = 7 * 24 * 3600

// NodeAdmission decides whether a registration may be applied. It is called by
// RegisterNodeGuarded UNDER the state lock, with the live node slice, and its
// answer is acted on before that lock is released.
//
// That is the whole point of the type. The pin it implements ("one credential,
// one node record") used to be a separate method that took and released the
// state lock on its own, so N concurrent registrations from one credential could
// all pass the check before any of them appended — a small permanent inflation
// of the node table per compromised credential. Check-and-insert now share one
// unbroken critical section, the same shape the tenant-quota TOCTOU was fixed
// with (see quotaExceededLocked, called under the lock that also inserts).
//
// It must not block, take any other lock, or call back into State.
type NodeAdmission func(nodes []*model.Node) (reason string, ok bool)

// NodeRegistration is the outcome of RegisterNodeGuarded.
type NodeRegistration struct {
	// ID is the registered node's id, or "" when nothing was registered.
	ID string
	// Refused is non-empty exactly when the admission check said no; it is the
	// reason, for the caller to log and return. An empty Refused with an empty
	// ID means the table was full and nothing could be reclaimed.
	Refused string
	// Reclaimed lists the ids of provably-stale records evicted to make room.
	// Normally empty: reclamation runs only at the ceiling.
	Reclaimed []string
}

// RegisterNode registers a node idempotently by hostname, with no admission
// check. Returns the node id, or "" if a NEW record would exceed MaxNodes (an
// existing node always re-registers, so a full list never stops the fleet from
// heartbeating).
func (s *State) RegisterNode(hostname, addr string, cpus uint32, memTotal uint64, now int64) string {
	return s.RegisterNodeGuarded(hostname, addr, cpus, memTotal, now, nil).ID
}

// RegisterNodeGuarded is RegisterNode with an admission check that runs under
// the SAME lock acquisition as the append it authorizes — see NodeAdmission for
// why that is not optional. A nil admit admits everything.
func (s *State) RegisterNodeGuarded(hostname, addr string, cpus uint32, memTotal uint64, now int64, admit NodeAdmission) NodeRegistration {
	s.mu.Lock()
	defer s.mu.Unlock()

	// Inside the critical section, and nothing between here and the append
	// below releases the lock: whatever admit saw is still true when the record
	// lands.
	if admit != nil {
		if reason, ok := admit(s.Nodes); !ok {
			return NodeRegistration{Refused: reason}
		}
	}

	if n := s.findNodeByHostname(hostname); n != nil {
		// Update mutable fields, keep id.
		n.Addr = addr
		n.CPUs = cpus
		n.MemTotalMiB = memTotal
		n.LastHB = now
		return NodeRegistration{ID: n.ID}
	}
	var reclaimed []string
	if len(s.Nodes) >= MaxNodes {
		// Only at the ceiling. In normal operation nothing is ever evicted, so
		// this cannot be a way to lose a node record by accident.
		reclaimed = s.reclaimStaleNodesLocked(now)
		if len(s.Nodes) >= MaxNodes {
			return NodeRegistration{Reclaimed: reclaimed}
		}
	}
	id := s.nextID("node")
	n := &model.Node{
		ID:          id,
		Hostname:    hostname,
		Addr:        addr,
		CPUs:        cpus,
		MemTotalMiB: memTotal,
		MemFreeMiB:  memTotal,
		VMCount:     0,
		PoolSize:    0,
		LastHB:      now,
	}
	s.Nodes = append(s.Nodes, n)
	return NodeRegistration{ID: id, Reclaimed: reclaimed}
}

// reclaimStaleNodesLocked drops node records that are PROVABLY dead, to make
// room at the MaxNodes ceiling. Caller must hold the lock.
//
// Two conditions, both required, and both about being sure rather than being
// thorough:
//
//   - silent for at least StaleNodeTTL. See that constant: it is ~120,000 missed
//     heartbeats, so a node that is merely rebooting is nowhere near it.
//   - no sandbox references the record. A node id is how hearthd finds the
//     address to reach a sandbox's microVM; dropping a record that still owns
//     workloads would strand them exactly the way an orphaned VM is stranded.
//     This is also what makes the eviction cheap to be wrong about — a reclaimed
//     record stands for nothing that is running.
//
// A node reclaimed in error is not lost: its agent's next register call creates
// a fresh record (registration is keyed on hostname), which is precisely the
// path that was 503ing before. A table full of stale-but-still-referenced
// records reclaims nothing and still 503s — the conservative direction.
func (s *State) reclaimStaleNodesLocked(now int64) []string {
	inUse := make(map[string]bool, len(s.Sandboxes))
	for _, sb := range s.Sandboxes {
		if sb.NodeID != nil {
			inUse[*sb.NodeID] = true
		}
	}
	var reclaimed []string
	total := len(s.Nodes)
	kept := s.Nodes[:0]
	for _, n := range s.Nodes {
		if now-n.LastHB >= StaleNodeTTL && !inUse[n.ID] {
			reclaimed = append(reclaimed, n.ID)
			continue
		}
		kept = append(kept, n)
	}
	// Clear the tail so the evicted records are not retained by the backing
	// array (the list is MaxNodes-sized by the time this runs).
	for i := len(kept); i < total; i++ {
		s.Nodes[i] = nil
	}
	s.Nodes = kept
	return reclaimed
}

// Heartbeat updates a node's live stats. Returns false if the node is unknown.
func (s *State) Heartbeat(id string, memFree uint64, vmCount, poolSize uint32, now int64) bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	n := s.FindNode(id)
	if n == nil {
		return false
	}
	n.MemFreeMiB = memFree
	n.VMCount = vmCount
	n.PoolSize = poolSize
	n.LastHB = now
	return true
}

// PickNode picks the ready node with the lowest vm_count.
// Caller must hold the lock.
func (s *State) PickNode(now int64) *model.Node {
	var best *model.Node
	for _, n := range s.Nodes {
		if n.NodeStatus(now) != "ready" {
			continue
		}
		if best == nil || n.VMCount < best.VMCount {
			best = n
		}
	}
	return best
}

// ---- Sandboxes ----

// FindSandbox returns the sandbox with the given id, or nil.
// Caller must hold the lock.
func (s *State) FindSandbox(id string) *model.Sandbox {
	for _, sb := range s.Sandboxes {
		if sb.ID == id {
			return sb
		}
	}
	return nil
}

// CreateSandbox creates a sandbox in "creating" state and appends it to the store.
// Caller must hold the lock.
func (s *State) CreateSandbox(name, namespace, nodeID string, vcpus uint32, memMiB uint64, now int64) *model.Sandbox {
	id := s.nextID("sb")
	nid := nodeID
	sb := &model.Sandbox{
		ID:        id,
		Name:      name,
		Namespace: namespace,
		NodeID:    &nid,
		State:     model.StateCreating,
		VCPUs:     vcpus,
		MemMiB:    memMiB,
		IP:        nil,
		CreatedAt: now,
		ParentID:  nil,
	}
	s.Sandboxes = append(s.Sandboxes, sb)
	return sb
}

// CreateForkChild creates a child sandbox in "creating" state.
// Caller must hold the lock.
func (s *State) CreateForkChild(name, namespace, nodeID string, vcpus uint32, memMiB uint64, parentID string, now int64) *model.Sandbox {
	id := s.nextID("sb")
	nid := nodeID
	pid := parentID
	sb := &model.Sandbox{
		ID:        id,
		Name:      name,
		Namespace: namespace,
		NodeID:    &nid,
		State:     model.StateCreating,
		VCPUs:     vcpus,
		MemMiB:    memMiB,
		IP:        nil,
		CreatedAt: now,
		ParentID:  &pid,
	}
	s.Sandboxes = append(s.Sandboxes, sb)
	return sb
}

// SetSandboxState updates a sandbox's state field.
func (s *State) SetSandboxState(id string, st model.SandboxState) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if sb := s.FindSandbox(id); sb != nil {
		sb.State = st
	}
}

// SetSandboxIP updates a sandbox's IP field. Empty string clears to nil.
func (s *State) SetSandboxIP(id, ip string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	sb := s.FindSandbox(id)
	if sb == nil {
		return
	}
	if ip == "" {
		sb.IP = nil
	} else {
		v := ip
		sb.IP = &v
	}
}

// RemoveSandbox removes a sandbox by id.
func (s *State) RemoveSandbox(id string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	for i, sb := range s.Sandboxes {
		if sb.ID == id {
			s.Sandboxes = append(s.Sandboxes[:i], s.Sandboxes[i+1:]...)
			return
		}
	}
}

// RecordWake updates wake metrics.
func (s *State) RecordWake(ms uint64) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.WakeMsLast = ms
	s.WakeTotal++
	s.WakeMsSum += ms
}

// RecordFork increments the fork counter.
func (s *State) RecordFork() {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.ForksTotal++
}

// RecordExec increments the exec attempt counter.
func (s *State) RecordExec() {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.ExecsTotal++
}

// ---- Persistence ----

// persistState is the on-disk JSON format, mirroring Zig's output exactly.
type persistState struct {
	Seq          uint64            `json:"seq"`
	RequestCount uint64            `json:"request_count"`
	Nodes        []json.RawMessage `json:"nodes"`
	Sandboxes    []*model.Sandbox  `json:"sandboxes"`
}

// Persist atomically writes state to path (temp file + rename).
// The node status field is computed with each node's own last_heartbeat as "now"
// so status is always "ready", matching Zig's persist behavior exactly.
func (s *State) Persist(path string) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	// Marshal nodes using last_heartbeat as "now" so status is always "ready"
	// (exactly what Zig does: n.writeJson(&b, n.last_heartbeat)).
	nodes := make([]json.RawMessage, len(s.Nodes))
	for i, n := range s.Nodes {
		b, err := n.MarshalWithNow(n.LastHB)
		if err != nil {
			return fmt.Errorf("marshal node: %w", err)
		}
		nodes[i] = json.RawMessage(b)
	}

	ps := persistState{
		Seq:          s.Seq,
		RequestCount: s.RequestCount,
		Nodes:        nodes,
		Sandboxes:    s.Sandboxes,
	}

	data, err := json.Marshal(ps)
	if err != nil {
		return fmt.Errorf("marshal state: %w", err)
	}

	// 0600, not 0644: this file is the whole fleet inventory — every node's
	// address, every sandbox id and the tenant each belongs to. hearthd is the
	// only reader and it runs as root, so nothing needs the group/other bits,
	// and the sibling persistence paths (store/sqlite.go's DB, wg.go's private
	// keys) are already 0600. The mode is on the temp file because the rename
	// carries it over: setting it afterwards would leave a world-readable
	// window at the real path on every write.
	tmp := path + ".tmp"
	if err := os.WriteFile(tmp, data, 0o600); err != nil {
		return fmt.Errorf("write tmp: %w", err)
	}
	if err := os.Rename(tmp, path); err != nil {
		return fmt.Errorf("rename: %w", err)
	}
	return nil
}

// loadNode is used when reading nodes from the state file.
type loadNode struct {
	ID          string `json:"id"`
	Hostname    string `json:"hostname"`
	Addr        string `json:"addr"`
	CPUs        uint32 `json:"cpus"`
	MemTotalMiB uint64 `json:"mem_total_mib"`
	MemFreeMiB  uint64 `json:"mem_free_mib"`
	VMCount     uint32 `json:"vm_count"`
	PoolSize    uint32 `json:"pool_size"`
	// "status" is ignored on load (computed at runtime).
	LastHB int64 `json:"last_heartbeat"`
}

// loadSandbox allows tolerating null/missing fields when reading from file.
type loadSandbox struct {
	ID        string  `json:"id"`
	Name      string  `json:"name"`
	Namespace string  `json:"namespace"`
	NodeID    *string `json:"node_id"`
	State     string  `json:"state"`
	VCPUs     uint32  `json:"vcpus"`
	MemMiB    uint64  `json:"mem_mib"`
	IP        *string `json:"ip"`
	CreatedAt int64   `json:"created_at"`
	ParentID  *string `json:"parent_id"`
	// Ingress (v4 P3) — Persist writes the model wholesale; reading must not
	// silently drop these (dead routes + orphaned DNAT rules on restart).
	Exposes           []model.Expose `json:"exposes"`
	AllowDynamicPorts bool           `json:"allow_dynamic_ports"`
	// Templates & disk (v4 P4) — same rule: dropping these on load would
	// corrupt disk-quota accounting after a restart.
	Template string `json:"template"`
	DiskGB   uint32 `json:"disk_gb"`
}

// loadFile is the shape used when reading the persisted JSON.
type loadFile struct {
	Seq          uint64            `json:"seq"`
	RequestCount uint64            `json:"request_count"`
	Nodes        []json.RawMessage `json:"nodes"`
	Sandboxes    []json.RawMessage `json:"sandboxes"`
}

// Load reads state from path. Missing file is not an error at the call site;
// the caller decides whether to warn. Unknown sandbox states map to "stopped".
func (s *State) Load(path string) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	data, err := os.ReadFile(path)
	if err != nil {
		return err
	}

	var lf loadFile
	if err := json.Unmarshal(data, &lf); err != nil {
		return fmt.Errorf("parse state: %w", err)
	}

	s.Seq = lf.Seq
	s.RequestCount = lf.RequestCount

	for _, raw := range lf.Nodes {
		var ln loadNode
		if err := json.Unmarshal(raw, &ln); err != nil {
			continue
		}
		if ln.ID == "" {
			continue
		}
		s.Nodes = append(s.Nodes, &model.Node{
			ID:          ln.ID,
			Hostname:    ln.Hostname,
			Addr:        ln.Addr,
			CPUs:        ln.CPUs,
			MemTotalMiB: ln.MemTotalMiB,
			MemFreeMiB:  ln.MemFreeMiB,
			VMCount:     ln.VMCount,
			PoolSize:    ln.PoolSize,
			LastHB:      ln.LastHB,
		})
	}

	for _, raw := range lf.Sandboxes {
		var ls loadSandbox
		if err := json.Unmarshal(raw, &ls); err != nil {
			continue
		}
		if ls.ID == "" {
			continue
		}
		// Default missing namespace.
		ns := ls.Namespace
		if ns == "" {
			ns = "default"
		}
		// Default missing vcpus/mem.
		vcpus := ls.VCPUs
		if vcpus == 0 {
			vcpus = 1
		}
		memMiB := ls.MemMiB
		if memMiB == 0 {
			memMiB = 256
		}
		sb := &model.Sandbox{
			ID:        ls.ID,
			Name:      ls.Name,
			Namespace: ns,
			NodeID:    ls.NodeID,
			State:     model.ParseSandboxState(ls.State),
			VCPUs:     vcpus,
			MemMiB:    memMiB,
			IP:        ls.IP,
			CreatedAt: ls.CreatedAt,
			ParentID:  ls.ParentID,
			Exposes:           ls.Exposes,
			AllowDynamicPorts: ls.AllowDynamicPorts,
			Template:          ls.Template,
			DiskGB:            ls.DiskGB,
		}
		s.Sandboxes = append(s.Sandboxes, sb)
	}

	return nil
}

// EnsureStateDir creates the directory for path if it doesn't exist.
func EnsureStateDir(path string) error {
	dir := filepath.Dir(path)
	return os.MkdirAll(dir, 0o755)
}
